This document demonstrates the analysis of Seurat objects. Those Seurat
Objects were derived from Dropest Count Matrices Using Different GTF
Files. Feature Representation column explains which transcript/ gene
were assigned as gene feature in the corresponding GTF file.

![table of nomenclature for different Seurat
objects](/data/gpfs/projects/punim2183/data_processed/table_naming_seurats.png)

## Load libraries

    library(Seurat)
    library(cowplot)
    library(dplyr)
    library(ggplot2)
    library(readr)
    library(DT)
    library(tidyr)
    library(patchwork)
    library(harmony)
    library(CSCORE)
    library(WGCNA)
    library(pheatmap)
    library(corrplot)
    library(rtracklayer)
    library(clusterProfiler)
    library(org.Mm.eg.db)
    library(enrichplot)
    library(glmGamPoi)


    source("/data/gpfs/projects/punim2183/data_processed/dropEst_analysis_functions.R")
    #save.image(file = "dropEst_output.RData")
    #load("dropEst_output.RData")

    knitr::opts_chunk$set(eval = FALSE)
     # 
     # knitr::knit("dropEst_output.Rmd", output = "dropEst_output_analysis.md")

     rmarkdown::render("dropEst_output.Rmd", output_format = "md_document", output_file = "dropEst_output_analysis.md")

The Following steps explain how the Dropest generated count matrices are
converted and saved as an rds Seurat objects:

## Step 1: Reading count matrices from a directory and createing Seurat objects:

For each count matrix type (trunc, FL, orig and both) generated by
dropest using different gtf file.

    # Define the directory containing RDS files
    rds_directory_both <- "/data/gpfs/projects/punim2183/samples_processing/count_matrix_files/both"

    # Step 1: Read RDS files into Seurat objects
    seurat_list_both <- read_rds_to_seurat(rds_directory_both)

# Step 2: Rename cells in the Seurat objects:

Extract naming patterns from the names of Seurat objects. And rename
cell names in each Seurat object using the extracted naming patterns.
Then combine multiple Seurat objects into a single Seurat object.

*Extract the naming pattern (e.g., “10X05\_1”)*

**input** seurat\_objects\_both (a list of Seurat objects with names
like count\_matrix.rds\_10X05\_1.bam.1\_both.gtf).

**Output** Each Seurat object in seurat\_objects\_both now has cell
names prefixed with the corresponding naming pattern (e.g.,
“10X05\_1\_Cell\_1”).

    seurat_list_both <- rename_cells_in_seurat_objects(seurat_list_both)

# Step 3: Normalize and merge Seurat objects

Normalise each seurat object then merge into one object.

**Input:** List of seurat objects .

**Output:** A Normalised and merged Seurat object with:

    combined_seurat_both <- normalize_and_merge(seurat_list_both)

# Step 4: Update orig.ident metadata

Update the orig.ident metadata column based on the sample names
extracted from the cell names.

**input:** merged seurat object, **output:** updated metadata with
correspondent cell names.

    combined_seurat_both <- update_orig_ident(combined_seurat_both)

# Step 5: Process the merged Seurat object

Process the combined Seurat object by scaling, clustering, and filtering
the data.

**Input:** combined\_seurat\_both (the combined Seurat object).

**Output:** A processed Seurat object with:

-   scaled data.

-   PCA (npc = 20), clustering, and UMAP performed.

-   Cells filtered based on mitochondrial content (percent.mt &lt; 5),
    feature count (nFeature\_RNA &gt; 200 & nFeature\_RNA &lt; 2000),
    and RNA count.

<!-- -->

    combined_seurat_both <- process_seurat(combined_seurat_both)

# Step 6: Integrate datasets using Harmony

For addressing batch effect, grouped by orig.ident

# Step 7: Save the final Seurat object

# Both Seurat

*Processing “Both” originated Seurat*

Check the representation of the transcript isoforms as genes in the
metadata

## Further processing: clustering and cell type annotation:

Create clusters and annotate cell-types to clusters using marker genes
of cell types of interest

Check reductions done to the object for clustering, and join assay
layers for Seurat v5

# Assign Clusters to cell types of interest

## Subset seurat to cell types of interest for focused plotting

# CS\_core/ gene co-expression

## gene co-expression analysis:

### Ntrk2trunc

### Ntrk2FL

# GO enrichment

Prepare Gene Lists:

brain\_gene\_list\_FL contains genes correlated with the full-length
isoform (Ntrk2FL) with a correlation coefficient **(r) &gt; 0.2**.

brain\_gene\_list\_trunc contains genes correlated with the truncated
isoform (Ntrk2trunc) with a correlation coefficient **(r) &gt; 0.2**.

\*Convert Gene Symbols to ENTREZ IDs: Gene symbols (SYMBOL) are
converted to ENTREZID format using the bitr function from
clusterProfiler. This is necessary because GO enrichment analysis
typically works with ENTREZID.

\*Run GO Enrichment for Each Ontology: The function is applied to both
gene lists (for full-length and truncated isoforms) across three
different GO ontologies: Biological Process (BP), Molecular Function
(MF), and Cellular Component (CC).

\*Generate Plots: A GO enrichment dot plot is generated for each
combination of gene list and ontology.

Biological Process (BP): GO terms related to biological functions and
processes in the cell or organism.

Molecular Function (MF): GO terms that describe the molecular activities
of the gene products.

Cellular Component (CC): GO terms related to the location of the gene
product in the cell.

To compare GO enrichment results between two gene clusters (FL and
trunc) across different ontology types (BP, MF, CC), and visualize GO
terms uniquely enriched in one cluster but not the other.

    # List of ontologies to loop through
    ontologies <- c("BP", "MF", "CC")

    # Function to generate plots for each ontology
    generate_plots_for_ontology <- function(ontology) {
      # Compare GO terms for each cluster (trunc and FL)
      compare_go <- compareCluster(
        geneCluster = gene_lists,
        fun = "enrichGO",
        OrgDb = org.Mm.eg.db,
        keyType = "ENTREZID",
        ont = ontology,
        pvalueCutoff = 0.05,
        pAdjustMethod = "BH",
        qvalueCutoff = 0.2,
        readable = TRUE
      )
      
      go_df <- compare_go@compareClusterResult
      print(paste("Number of GO terms for ontology", ontology, ":", nrow(go_df)))  # Debugging
      
      # Filter significant results
      filtered_go <- go_df %>%
        filter(p.adjust < 0.05) %>%
        dplyr::select(Description, Cluster, p.adjust, GeneRatio)
      
      print(paste("Number of significant GO terms for ontology", ontology, ":", nrow(filtered_go)))  # Debugging
      
      # If there are no significant terms, stop the function here
      if (nrow(filtered_go) == 0) {
        print(paste("No significant GO terms for ontology", ontology))
        return(NULL)  # Exit if no significant results
      }
      
      # Separate by Cluster (FL vs trunc) and examine the differences
      fl_terms <- filtered_go %>% filter(Cluster == "FL")
      trunc_terms <- filtered_go %>% filter(Cluster == "trunc")
      
      # Identify the GO terms that are different (unique for each isoform)
      fl_unique <- setdiff(fl_terms$Description, trunc_terms$Description)
      trunc_unique <- setdiff(trunc_terms$Description, fl_terms$Description)
      
      # Select top N unique terms by significance (lowest p.adjust)

      top_n <- 15

      fl_top <- fl_terms %>%
      filter(Description %in% fl_unique) %>%
      arrange(p.adjust) %>%
      slice_head(n = top_n)

      trunc_top <- trunc_terms %>%
      filter(Description %in% trunc_unique) %>%
      arrange(p.adjust) %>%
      slice_head(n = top_n)

      
     top_unique_terms <- data.frame(
      Term = c(fl_top$Description, trunc_top$Description),
      Cluster = c(rep("FL", nrow(fl_top)), rep("trunc", nrow(trunc_top))),
      stringsAsFactors = FALSE
    )
     
      # If there are no unique terms, stop the function here
      if (nrow(top_unique_terms) == 0) {
        print(paste("No unique GO terms for ontology", ontology))
        return(NULL)  # Exit if no unique terms
      }
      
      # Plot using ggplot
      p <- ggplot(top_unique_terms, aes(x = Term, y = Cluster, color = Cluster)) +
        geom_point(size = 4) +
        coord_flip() +
        labs(title = paste("Top Unique GO Terms:", ontology, "- Ntrk2FL vs Ntrk2trunc"),
             x = "GO Term", y = "Cluster") +
        scale_color_manual(values = c("blue", "red")) +  
        theme_minimal() +
        theme(plot.title = element_text(hjust = 0),  
              plot.title.position = "plot",          
              axis.text.x = element_text(angle = 45, hjust = 1),
              axis.text.y = element_text(size = 10))
      print(p)
      
      ggsave(
      filename = paste0("GO_TopUnique_", ontology, ".png"),
      plot = p,
      width = 10,
      height = 6,
      dpi = 300
    )
    }
    # Generate plots for each ontology
    for (ontology in ontologies) {
      generate_plots_for_ontology(ontology)
    }

# M10 Seurat

*Processing M10 originated Seurat*

    # Define the directory containing RDS files
    rds_directory_orig <- "/data/gpfs/projects/punim2183/samples_processing/count_matrix_files/orig"

    # 1. Load with renaming data
    seurat_list_M10 <- read_rds_to_seurat(rds_directory_orig) |>
      rename_cells_in_seurat_objects()

    # 2. Normalize with SCTransform and merge
    combined_seurat_M10 <- normalize_and_merge(seurat_list_M10) |>
      update_orig_ident()

    # 3. Process with Harmony integration
    combined_seurat_M10 <- process_seurat(combined_seurat_M10)
    combined_seurat_M10 <- run_harmony(combined_seurat_M10)

    # 4. Save the processed object
    saveRDS(combined_seurat_M10, "combined_seurat_M10.rds")


    # Print a success message
    print("Workflow completed successfully! The combined Seurat object has been saved.")

    #saving files and loading the previous step to save computational time
    #combined_seurat_both <- readRDS("/data/gpfs/projects/punim2183/data_processed/combined_seurat_both.rds")

    names(combined_seurat_M10@reductions)

    combined_seurat_M10 <- JoinLayers(combined_seurat_M10, assay = "RNA")

    cluster_to_celltype <- list(
      Neurons = c(5, 3, 20, 28, 19, 11, 14),
      Oligo = c(10, 16, 17, 31, 7, 27, 29),
      Astrocytes = c(1, 2, 8, 18, 9, 26, 23, 29)
    )

    # Create a vector to store cell type labels
    cell_type_labels_M10 <- rep(NA, ncol(combined_seurat_M10))

    # Assign specific cell types based on cluster IDs
    for (cluster_id in 0:31) {
      if (cluster_id %in% cluster_to_celltype$Neurons) {
        cell_type_labels_M10[combined_seurat_M10$seurat_clusters == cluster_id] <- "Neurons"
      } else if (cluster_id %in% cluster_to_celltype$Oligo) {
        cell_type_labels_M10[combined_seurat_M10$seurat_clusters == cluster_id] <- "Oligo"
      } else if (cluster_id %in% cluster_to_celltype$Astrocytes) {
        cell_type_labels_M10[combined_seurat_M10$seurat_clusters == cluster_id] <- "Astrocytes"
      }
    }

    # Assign generic labels to remaining clusters
    remaining_clusters_M10 <- setdiff(0:31, unlist(cluster_to_celltype))
    for (i in seq_along(remaining_clusters_M10)) {
      cluster_id <- remaining_clusters_M10[i]
      cell_type_labels_M10[combined_seurat_M10$seurat_clusters == cluster_id] <- paste0("cell_type", i)
    }


    # Add the cell type labels to the Seurat object
    combined_seurat_M10$cell_type <- cell_type_labels_M10

    # Set the cell type as the active identity
    Idents(combined_seurat_M10) <- combined_seurat_M10$cell_type
    table(combined_seurat_M10$cell_type)

    dimplot_cell_types_M10 <- DimPlot(combined_seurat_M10, reduction = "umap", label = TRUE, group.by = "cell_type", raster = FALSE)
    dimplot_cell_types_M10
    ggsave("dimplot_cell_types_M10.jpg", plot= dimplot_cell_types_M10,width = 10, height = 8, dpi = 300)

    #check metadata
    M10_metadata <- combined_seurat_M10@meta.data
    head(M10_metadata)

## Subset seurat to cell types of interest for focused plotting

\`\`\`
