This document demonstrates the analysis of Seurat objects. Those Seurat Objects were derived from Dropest Count Matrices Using Different GTF Files. Feature Representation column explains which transcript/ gene were assigned as gene feature in the corresponding GTF file.

![table of nomenclature for different Seurat objects](/data/gpfs/projects/punim2183/data_processed/table_naming_seurats.png)



## Load libraries

```{r}
library(Seurat)
library(cowplot)
library(dplyr)
library(ggplot2)
library(readr)
library(DT)
library(tidyr)
library(patchwork)
library(harmony)
library(CSCORE)
library(WGCNA)
library(pheatmap)
library(corrplot)
library(rtracklayer)
library(clusterProfiler)
library(org.Mm.eg.db)
library(enrichplot)
library(glmGamPoi)
library(stringr)

library(SingleR)
library(celldex)
library(Seurat)
library(SingleCellExperiment)


source("/data/gpfs/projects/punim2183/data_processed/dropEst_analysis_functions.R")
#save.image(file = "dropEst_output.RData")
#load("dropEst_output.RData")

knitr::opts_chunk$set(eval = FALSE)
 # 
 # knitr::knit("dropEst_output.Rmd", output = "dropEst_output_analysis.md")

 #rmarkdown::render("dropEst_output.Rmd", output_format = "md_document", output_file = "dropEst_output_analysis.md")
```

The Following steps explain how the Dropest generated count matrices are converted and saved as an rds Seurat objects: 

## Step 1: Reading count matrices from a directory and createing Seurat objects:

For each count matrix type (trunc, FL, orig and both) generated by dropest using different gtf file.

```{r}

# Define the directory containing RDS files
rds_directory_both <- "/data/gpfs/projects/punim2183/samples_processing/count_matrix_files/both"

# Step 1: Read RDS files into Seurat objects
seurat_list_both <- read_rds_to_seurat(rds_directory_both)
```

# Step 2: Rename cells in the Seurat objects:

Extract naming patterns from the names of Seurat objects. And rename cell names in each Seurat object using the extracted naming patterns. Then combine multiple Seurat objects into a single Seurat object.

*Extract the naming pattern (e.g., "10X05_1")* 

**input** combined_seurat_bothects_both (a list of Seurat objects with names like count_matrix.rds_10X05_1.bam.1_both.gtf).


**Output** Each Seurat object in combined_seurat_bothects_both now has cell names prefixed with the corresponding naming pattern (e.g., "10X05_1\_Cell_1").

```{r}
seurat_list_both <- rename_cells_in_combined_seurat_bothects(seurat_list_both)
```

# Step 3: Normalize and merge Seurat objects


Normalise each seurat object then merge into one object. 

**Input:** List of seurat objects .

**Output:** A Normalised and merged Seurat object with:

```{r}
combined_seurat_both <- normalize_and_merge(seurat_list_both)
```

# Step 4: Update orig.ident metadata
 Update the orig.ident metadata column based on the sample names extracted from the cell names.

 **input:** merged seurat object, 
 **output:** updated metadata with correspondent cell names.
```{r}
combined_seurat_both <- update_orig_ident(combined_seurat_both)
```

# Step 5: Process the merged Seurat object

Process the combined Seurat object by scaling, clustering, and filtering the data. 

**Input:** combined_seurat_both (the combined Seurat object).

**Output:** A processed Seurat object with:

-   scaled data.

-   PCA (npc = 20), clustering, and UMAP performed.

-   Cells filtered based on mitochondrial content (percent.mt \< 5), feature count (nFeature_RNA \> 200 & nFeature_RNA \< 2000), and RNA count.

```{r}
combined_seurat_both <- process_seurat(combined_seurat_both)
```

# Step 6: Integrate datasets using Harmony

For addressing batch effect, grouped by orig.ident

```{r, echo=FALSE}
combined_seurat_both <- run_harmony(combined_seurat_both)
dimplot_orig_ident <- DimPlot(combined_seurat_both, reduction = "umap", group.by = "orig.ident") 

ggsave("dimplot_orig_ident.png", plot = dimplot_orig_ident, width = 10, height = 8, dpi = 300)
```

# Step 7: Save the final Seurat object

```{r, echo=FALSE}
save_combined_seurat_bothect(combined_seurat_both, "combined_seurat_both.rds")

# Print a success message
print("Workflow completed successfully! The combined Seurat object has been saved.")
```




# Both Seurat

*Processing "Both" originated Seurat*

```{r, echo=FALSE}
# Define the directory containing RDS files
rds_directory_both <- "/data/gpfs/projects/punim2183/samples_processing/count_matrix_files/both/cm_both/"

# 1. Load with renaming data
seurat_list_both <- read_rds_to_seurat(rds_directory_both) |> rename_cells_in_seurat_objects()

seurat_list_both <- readRDS("/data/gpfs/projects/punim2183/data_processed/seurat_list_both.rds")
# 2. Normalize and merge
combined_seurat_both <- normalize_and_merge(seurat_list_both)

combined_seurat_both <- update_orig_ident(combined_seurat_both)

# 3. Process with Harmony integration
combined_seurat_both <- process_seurat(combined_seurat_both)
combined_seurat_both <- run_harmony(combined_seurat_both, group.by.vars = "orig.ident", npcs = 20)

# 4. Save the processed object
saveRDS(combined_seurat_both, "combined_seurat_both.rds")

# Print a success message
print("Workflow completed successfully! The combined Seurat object has been saved.")


```



Check the representation of the transcript isoforms as genes in the metadata
```{r, echo=FALSE}

feature_plots <- FeaturePlot(combined_seurat_both, features = c("Ntrk2trunc", "Ntrk2FL"), raster = FALSE)
feature_plots[[1]] <- feature_plots[[1]] + labs(title = "TrkB.T1")
feature_plots[[2]] <- feature_plots[[2]] + labs(title = "TrkB.FL")
feature_plots
ggsave("isoforms_feature_plot.png", plot = feature_plots, width = 10, height = 8, dpi = 300)

```


## Further processing: clustering and cell type annotation: 
Create clusters and annotate cell-types to clusters using marker genes of cell types of interest

Check reductions done to the object for clustering, and join assay layers for Seurat v5
```{r, echo=FALSE}
#saving files and loading the previous step to save computational time
#combined_seurat_both <- readRDS("/data/gpfs/projects/punim2183/data_processed/combined_seurat_both.rds")

names(combined_seurat_both@reductions)

combined_seurat_both <- JoinLayers(combined_seurat_both, assay = "RNA")

```

```{r, fig.width=12, fig.height=10, echo=FALSE}
 combined_seurat_both <- readRDS("/data/gpfs/projects/punim2183/data_processed/combined_seurat_both.rds")

neurons_markers = c("Map2", "Rbfox3", "Tubb3")
oligos_markers = c("Mbp", "Olig1", "Olig2")
astrocytes_markers = c("Gfap", "Aqp4", "Slc1a3")

count_expressing_cells <- function(markers, combined_seurat_both) {
  expression_data <- FetchData(combined_seurat_both, vars = markers)
  num_cells <- sum(rowSums(expression_data > 0) > 0)
  return(num_cells)
}

markers_feature_plot <- FeaturePlot(combined_seurat_both, c(neurons_markers, astrocytes_markers, oligos_markers), raster = FALSE)
markers_feature_plot
#ggsave("markers_feature_plot.png", plot = markers_feature_plot, width = 10, height = 8, dpi = 300)


```

# Cell annotation:

```{r}
combined_seurat_both <- readRDS("/data/gpfs/projects/punim2183/data_processed/combined_seurat_both.rds")

combined_seurat_both <- JoinLayers(combined_seurat_both, assay = "RNA")

ref <- celldex::MouseRNAseqData()

sce <- as.SingleCellExperiment(combined_seurat_both, assay = "RNA")

#pred <- SingleR(test = sce, ref = ref, labels = ref$label.main)

#combined_seurat_both$SingleR_labels <- pred$labels

#using this instead cause we've already clustered
cluster.pred <- SingleR(test = sce, ref = ref, labels = ref$label.main, clusters = combined_seurat_both$seurat_clusters)
combined_seurat_both$SingleR_cluster_labels <- cluster.pred$labels[combined_seurat_both$seurat_clusters]

cell_types_dim <- DimPlot(combined_seurat_both, group.by = "SingleR_cluster_labels", label = TRUE) +labs(title = "Cell Type Clusters")
cell_types_dim

saveRDS(combined_seurat_both, "combined_seurat_both.rds")
ggsave("cell_types_dim.png", plot = cell_types_dim, width = 10, height = 8, dpi = 300)

VlnPlot(combined_seurat_both, features = c("Ntrk2FL", "Ntrk2trunc"), group.by = "SingleR_cluster_labels", pt.size = FALSE)

library(Seurat)
library(ggplot2)
library(patchwork)

vln_FL <- VlnPlot(
  combined_seurat_both,
  features = "Ntrk2FL",
  group.by = "SingleR_cluster_labels",
  pt.size = FALSE
) + ggtitle("TrkB.FL")

vln_T1 <- VlnPlot(
  combined_seurat_both,
  features = "Ntrk2trunc",
  group.by = "SingleR_cluster_labels",
  pt.size = FALSE
) + ggtitle("TrkB.T1")

vln_FL 
vln_T1
ggsave("vln_FL.png", plot = vln_FL, width = 10, height = 8, dpi = 300)
ggsave("vln_T1.png", plot = vln_T1, width = 10, height = 8, dpi = 300)

# Replace with your gene names
gene1 <- "Ntrk2FL"
gene2 <- "Ntrk2trunc"

# Extract expression values from the normalized data slot
expr_values <- FetchData(combined_seurat_both, vars = c(gene1, gene2))
head(expr_values)

#each dot here is one cell showing how much of each isoform it expresses 
#There are clear diagonal streaks of dots, showing a positive correlation between TrkB.FL and TrkB.T1 expression in many cellsâ€”especially in neurons (purple) and astrocytes (red).

ggplot(expr_values, aes_string(x = gene1, y = gene2, color = "combined_seurat_both$SingleR_cluster_labels")) +
  geom_point(alpha = 0.5) +
  labs(
    title = "Expression of TrkB.T1 vs TrkB.FL by Cell Type",
    x = "TrkB.FL",
    y = "TrkB.T1",
    color = "Cell Type"  
  ) +
  theme_minimal()


```
# Assign Clusters to cell types of interest manually

```{r, echo=FALSE}


cluster_to_celltype <- list(
Neurons = c(0, 1, 4, 14, 11, 26, 13, 12),
Oligo = c(6,24,22,17,16,27),
Astrocytes = c(15, 8, 3, 23, 25)
)
# Create a vector to store cell type labels
cell_type_labels <- rep(NA, ncol(combined_seurat_both))

# Assign specific cell types based on cluster IDs
for (cluster_id in 0:29) {
  if (cluster_id %in% cluster_to_celltype$Neurons) {
    cell_type_labels[combined_seurat_both$seurat_clusters == cluster_id] <- "Neurons"
  } else if (cluster_id %in% cluster_to_celltype$Oligo) {
    cell_type_labels[combined_seurat_both$seurat_clusters == cluster_id] <- "Oligo"
  } else if (cluster_id %in% cluster_to_celltype$Astrocytes) {
    cell_type_labels[combined_seurat_both$seurat_clusters == cluster_id] <- "Astrocytes"
  }
}

# Assign generic labels to remaining clusters
remaining_clusters <- setdiff(0:29, unlist(cluster_to_celltype))
for (i in seq_along(remaining_clusters)) {
  cell_type_labels[combined_seurat_both$seurat_clusters == remaining_clusters[i]] <- paste0("cell_type", i)
}

# Add the cell type labels to the Seurat object
combined_seurat_both$cell_type <- cell_type_labels

# Set the cell type as the active identity
Idents(combined_seurat_both) <- combined_seurat_both$cell_type
table(combined_seurat_both$cell_type)

dimplot_cell_types <- DimPlot(combined_seurat_both, reduction = "umap", label = TRUE, group.by = "cell_type", raster = FALSE)
dimplot_cell_types
ggsave("dimplot_cell_types.jpg", plot= dimplot_cell_types,width = 10, height = 8, dpi = 300)

#check metadata
Both_metadata <- combined_seurat_both@meta.data
head(Both_metadata)


```


## Subset seurat to cell types of interest for focused plotting 
```{r, echo=FALSE}
cell_types_of_interest <- c("Neurons", "Oligo", "Astrocytes")

subset_seurat_both <- subset(combined_seurat_both, subset = cell_type %in% cell_types_of_interest)


# Generate Plots
vln_plot <- VlnPlot(subset_seurat_both, features = c("Ntrk2FL", "Ntrk2trunc"), pt.size = 0)
feature_plot_both <- FeaturePlot(subset_seurat_both, features = c("Ntrk2FL", "Ntrk2trunc"), raster = FALSE)
feature_plot_trunc <- FeaturePlot(subset_seurat_both, features = "Ntrk2trunc", raster = FALSE)
feature_plot_fl <- FeaturePlot(subset_seurat_both, features = "Ntrk2FL", raster = FALSE)
dim_plot <- DimPlot(subset_seurat_both)

ggsave("feature_plot_both.png", plot = feature_plot_both, width = 10, height = 8, dpi = 300)
ggsave("feature_plot_trunc.png", plot = feature_plot_trunc, width = 10, height = 8, dpi = 300)
ggsave("feature_plot_fl.png", plot = feature_plot_fl, width = 10, height = 8, dpi = 300)
ggsave("dim_plot.png", plot = dim_plot, width = 10, height = 8, dpi = 300)
ggsave("vln_plot.png", plot = vln_plot, width = 10, height = 8, dpi = 300)

#saveRDS(subset_seurat_both, "subset_seurat_both.rds")
```




# CS_core/ gene co-expression
```{r, echo=FALSE}
#subset_seurat_both <- readRDS("/data/gpfs/projects/punim2183/data_processed/subset_seurat_both.rds")

# get layer data (returns matrix) this is for Seurat V5(when dealing with the whole data not only the subset)
counts_matrix <- LayerData(combined_seurat_both, assay = "RNA", layer = "counts")
data_matrix <- LayerData(combined_seurat_both, assay = "RNA", layer = "data")

# set layer data
combined_seurat_both <- SetAssayData(
  object = combined_seurat_both,
  assay = "RNA",
  layer = "counts",
  new.data = as(counts_matrix, "dgCMatrix")  # Convert to sparse
)

# For normalized data
combined_seurat_both <- SetAssayData(
  object = combined_seurat_both,
  assay = "RNA",
  layer = "data",
  new.data = as(data_matrix, "dgCMatrix")
)


# Before running CSCORE, ensure data stays sparse
library(Matrix)
combined_seurat_both <- SetAssayData(
  combined_seurat_both,
  assay = "RNA",
  layer = "data",
  new.data = as(LayerData(combined_seurat_both, layer = "data"), "dgCMatrix")
)

coexp_results <- run_CSCORE_on_Seurat_v5(
  combined_seurat_both,
  assay = "RNA",
  layer = "data",
  n_genes = 1500, #top 1500 with meaningful expression
)

# coexp_results <- run_CSCORE_on_Seurat_v5_optimized(
#   combined_seurat_bothect = combined_seurat_both,
#   assay = "RNA",
#   layer = "data",
#   n_genes = 5000,
# )


coexp_matrix <- coexp_results$coexpression_matrix


# Save co-expression matrix as a CSV file
write.csv(coexp_matrix, "coexp_matrix.csv", row.names = TRUE)
# Save co-expression matrix as an RDS file
saveRDS(coexp_matrix, "coexp_matrix.rds")

coexp_matrix <- readRDS("coexp_matrix.rds")

```



## gene co-expression analysis:
### Ntrk2trunc

```{r, echo=FALSE}

target_gene <- "Ntrk2trunc"

target_gene_expression_trunc <- coexp_matrix[target_gene, ]

# Calculate correlations
# correlations <- apply(coexp_matrix, 1, function(gene_expression) {
#   cor(target_gene_expression_trunc, gene_expression, method = "pearson")
# })

correlation_df_trunc <- data.frame(
  gene = rownames(coexp_matrix),
  correlation = coexp_matrix["Ntrk2trunc", ],
  row.names = NULL
)

# Sort by absolute correlation
correlation_df_trunc <- correlation_df_trunc[order(-(correlation_df_trunc$correlation)), ]

# Filter for significant correlations
significant_correlations_trunc <- correlation_df_trunc[(correlation_df_trunc$correlation) > 0.4, ] 
sig_genes_trunc <- significant_correlations_trunc$gene


genes_0.1_corr_trunc <- correlation_df_trunc[(correlation_df_trunc$correlation) > 0.1, ]
genes_0.2_corr_trunc <- correlation_df_trunc[(correlation_df_trunc$correlation) > 0.2, ]


# Plot the top 20 correlated genes (barplot)
top_20_genes_trunc <- head(significant_correlations_trunc, 20)
top_20_genes_barplot_trunc <- ggplot(top_20_genes_trunc, aes(x = reorder(gene, correlation), y = correlation)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  coord_flip() +
  labs(title = paste("Top Genes Correlated with TrkB.T1"),
       x = "Gene",
       y = "Correlation") +
  theme_minimal()
top_20_genes_barplot_trunc

top_20_genes_trunc$gene
ggsave("top_20_genes_barplot_trunc.png", plot = top_20_genes_barplot_trunc, width = 10, height = 8, dpi = 300)

# Extract the top 10 genes
top20_genes_trunc <- head((significant_correlations_trunc$gene), 21)
top50_genes_trunc <- head((significant_correlations_trunc$gene), 51)

# Subset the expression matrix
heatmap_matrix_top20_trunc <- coexp_matrix[top20_genes_trunc, top20_genes_trunc]

# Check the dimensions of the heatmap matrix
dim(heatmap_matrix_top20_trunc)

# Plot heatmap

library(ComplexHeatmap)
library(circlize)
library(grid)

col_fun <- colorRamp2(c(0, 0.5 , 1), c("blue", "yellow", "red"))

col_labels <- colnames(heatmap_matrix_top20_trunc)
row_labels <- rownames(heatmap_matrix_top20_trunc)

# Define styles for column names
row_font_colors_trunc <- ifelse(row_labels == "Ntrk2trunc", "red", "black")
col_font_colors_trunc <- ifelse(col_labels == "Ntrk2trunc", "red", "black")
col_font_face_trunc <- ifelse(col_labels == "Ntrk2trunc", "bold", "plain")

heatmap_plot_trunc <- Heatmap(heatmap_matrix_top20_trunc,
        name = "Correlation",
        col = col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        row_names_gp = gpar(fontsize = 10, fontface = col_font_face_trunc, col = row_font_colors_trunc),
        column_names_gp = gpar(fontsize = 10, fontface = col_font_face_trunc, col = col_font_colors_trunc),
        column_names_rot = 45,
        rect_gp = gpar(col = "white", lwd = 1))


# Draw the heatmap
draw(heatmap_plot_trunc, 
     column_title= "Top 20 Genes Co-expressed with TrkB.T1",
   column_title_gp=grid::gpar(fontsize=16))

png("Ntrk2trunc_correlation_heatmap.png", width = 10, height = 8, units = "in", res = 300)  
draw(heatmap_plot_trunc,
     column_title = "Top 20 Genes Co-expressed with TrkB.T1",
     column_title_gp = gpar(fontsize = 16, fontface = "bold", col = "black"))
dev.off()  # Close the PNG device

```

### Ntrk2FL
```{r, echo=FALSE}
target_gene <- "Ntrk2FL"

target_gene_expression_FL <- coexp_matrix[target_gene, ]

# Calculate correlations
# correlations <- apply(coexp_matrix, 1, function(gene_expression) {
#   cor(target_gene_expression_FL, gene_expression, method = "pearson")
# })

correlation_df_FL <- data.frame(
  gene = rownames(coexp_matrix),
  correlation = coexp_matrix["Ntrk2FL", ],
  row.names = NULL
)

# Sort by absolute correlation
correlation_df_FL <- correlation_df_FL[order(-(correlation_df_FL$correlation)), ]

# Filter for significant correlations
significant_correlations_FL <- correlation_df_FL[(correlation_df_FL$correlation) > 0.4, ] 
sig_genes_FL <- significant_correlations_FL$gene


genes_0.1_corr_FL <- correlation_df_FL[(correlation_df_FL$correlation) > 0.1, ]
genes_0.2_corr_FL <- correlation_df_FL[(correlation_df_FL$correlation) > 0.2, ]

# Plot the top 20 correlated genes
top_genes <- head(significant_correlations_FL, 20)
top_20_genes_barplot_FL<- ggplot(top_genes, aes(x = reorder(gene, correlation), y = correlation)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  coord_flip() +
  labs(title = paste("Top Genes Correlated with TrkB.FL"),
       x = "Gene",
       y = "Correlation") +
  theme_minimal()

top_genes$gene


# Extract the top 10 genes
top20_genes_FL <- head((significant_correlations_FL$gene), 21)

# Subset the expression matrix
heatmap_matrix_top20_FL <- coexp_matrix[top20_genes_FL, top20_genes_FL]

# Check the dimensions of the heatmap matrix
dim(heatmap_matrix_top20_FL)

# Plot heatmap

library(ComplexHeatmap)
library(circlize)
library(grid)

col_labels <- colnames(heatmap_matrix_top20_FL)
row_labels <- rownames(heatmap_matrix_top20_FL)

# Define styles for column names
row_font_colors_FL <- ifelse(row_labels == "Ntrk2FL", "red", "black")
col_font_colors_FL <- ifelse(col_labels == "Ntrk2FL", "red", "black")
col_font_face_FL <- ifelse(col_labels == "Ntrk2FL", "bold", "plain")


heatmap_plot_FL <- Heatmap(heatmap_matrix_top20_FL,
        name = "Correlation",
        col = col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        row_names_gp = gpar(fontsize = 10, fontface = col_font_face_FL, col = row_font_colors_FL),
        column_names_gp = gpar(fontsize = 10, fontface = col_font_face_FL, col = col_font_colors_FL),
        column_names_rot = 45,
        rect_gp = gpar(col = "white", lwd = 1))


# Draw the heatmap
draw(heatmap_plot_FL, 
     column_title= "Top 20 Genes Co-expressed with TrkB.FL",
   column_title_gp=grid::gpar(fontsize=16))

png("Ntrk2FL_correlation_heatmap.png", width = 10, height = 8, units = "in", res = 300)  
draw(heatmap_plot_FL,
     column_title = "Top 10 Genes Co-expressed with TrkB.FL",
     column_title_gp = gpar(fontsize = 16, fontface = "bold", col = "black"))
dev.off()  # Close the PNG device

ggsave("top_20_genes_barplot_FL.png", plot = top_20_genes_barplot_FL, width = 10, height = 8, dpi = 300)



```



# GO enrichment
Prepare Gene Lists:

brain_gene_list_FL contains genes correlated with the full-length isoform (Ntrk2FL) with a correlation coefficient **(r) > 0.2**.

brain_gene_list_trunc contains genes correlated with the truncated isoform (Ntrk2trunc) with a correlation coefficient **(r) > 0.2**.

*Convert Gene Symbols to ENTREZ IDs:
Gene symbols (SYMBOL) are converted to ENTREZID format using the bitr function from clusterProfiler. This is necessary because GO enrichment analysis typically works with ENTREZID.

*Run GO Enrichment for Each Ontology:
The function is applied to both gene lists (for full-length and truncated isoforms) across three different GO ontologies: Biological Process (BP), Molecular Function (MF), and Cellular Component (CC).

*Generate Plots:
A GO enrichment dot plot is generated for each combination of gene list and ontology.

Biological Process (BP): GO terms related to biological functions and processes in the cell or organism.

Molecular Function (MF): GO terms that describe the molecular activities of the gene products.

Cellular Component (CC): GO terms related to the location of the gene product in the cell.

```{r, fig.width=10, fig.height=8, echo=FALSE}

brain_gene_list_FL <- genes_0.2_corr_FL$gene #correlated genes with Ntrk2 truncated isoform with r >0.2
brain_gene_list_trunc <- genes_0.2_corr_trunc$gene # correlated genes with full-length isoform with r > 0.2

#convert SYMBOL to ENTERZID 
converted_genes_trunc <- bitr(brain_gene_list_trunc, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)
brain_gene_list_trunc <- converted_genes_trunc$ENTREZID

converted_genes_FL <-bitr(brain_gene_list_FL, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)
brain_gene_list_FL <- converted_genes_FL$ENTREZID


# Run GO enrichment with converted gene IDs
go_plot_FL_cc <- plot_GO_enrichment(brain_gene_list_FL, gene_id_type = 'ENTREZID', 
                                 ontology = 'CC', pval_cutoff = 0.05, OrgDb = 'org.Mm.eg.db'+
                                 ggtitle("Gene Ontology Enrichment for TrkB.FL"))

go_plot_trunc_cc <- plot_GO_enrichment(brain_gene_list_trunc, gene_id_type = 'ENTREZID', 
                                    ontology = 'CC', pval_cutoff = 0.05, OrgDb = 'org.Mm.eg.db'+
                                    ggtitle("Gene Ontology Enrichment for Ntrk2trunc"))

go_plot_FL_mf <- plot_GO_enrichment(brain_gene_list_FL, gene_id_type = 'ENTREZID', 
                                 ontology = 'MF', pval_cutoff = 0.05, OrgDb = 'org.Mm.eg.db'+
                                 ggtitle("Gene Ontology Enrichment for Ntrk2FL"))

go_plot_trunc_mf <- plot_GO_enrichment(brain_gene_list_trunc, gene_id_type = 'ENTREZID', 
                                    ontology = 'MF', pval_cutoff = 0.05, OrgDb = 'org.Mm.eg.db'+
                                    ggtitle("Gene Ontology Enrichment for Ntrk2trunc"))

go_plot_FL_bp <- plot_GO_enrichment(brain_gene_list_FL, gene_id_type = 'ENTREZID', 
                                 ontology = 'BP', pval_cutoff = 0.05, OrgDb = 'org.Mm.eg.db'+
                                 ggtitle("Gene Ontology Enrichment for Ntrk2FL"))

go_plot_trunc_bp <- plot_GO_enrichment(brain_gene_list_trunc, gene_id_type = 'ENTREZID', 
                                    ontology = 'BP', pval_cutoff = 0.05, OrgDb = 'org.Mm.eg.db'+
                                    ggtitle("Gene Ontology Enrichment for Ntrk2trunc"))


# ggsave("go_plot_trunc_bp.png", plot = go_plot_trunc_bp, dpi = 300)
# ggsave("go_plot_FL_bp.png", plot = go_plot_FL_bp, dpi = 300)
# 
# ggsave("go_plot_trunc_mf.png", plot = go_plot_trunc_mf, dpi = 300)
# ggsave("go_plot_FL_mf.png", plot = go_plot_FL_mf, dpi = 300)
# 
# ggsave("go_plot_trunc_cc.png", plot = go_plot_trunc_cc, dpi = 300)
# ggsave("go_plot_FL_cc.png", plot = go_plot_FL_cc, dpi = 300)

```



To compare GO enrichment results between two gene clusters (FL and trunc) across different ontology types (BP, MF, CC), and visualize GO terms uniquely enriched in one cluster but not the other.

compareCluster():
- Takes a named list of gene vectors (in gene_lists, one for FL and one for trunc).

- Runs enrichGO() for each group.

- Returns combined results in one object (compare_go), making it easy to compare.

```{r, fig.width= 8, fig.height= 6}
# List of ontologies to loop through
ontologies <- c("BP", "CC", "MF")

gene_lists <- list(FL= converted_genes_FL$ENTREZID,
                   trunc = converted_genes_trunc$ENTREZID)

# Function to generate plots for each ontology
generate_plots_for_ontology <- function(ontology) {
  ontology_full <- switch(ontology,
  "BP" = "Biological Process",
  "CC" = "Cellular Component",
  "MF" = "Molecular Function",
  ontology  # default fallback
)
  # Compare GO terms for each cluster (trunc and FL)
  compare_go <- compareCluster(
    geneCluster = gene_lists,
    fun = "enrichGO",
    OrgDb = org.Mm.eg.db,
    keyType = "ENTREZID",
    ont = ontology,
    pvalueCutoff = 0.05,
    pAdjustMethod = "BH",
    qvalueCutoff = 0.2,
    readable = TRUE
  )
  
  go_df <- compare_go@compareClusterResult
  print(paste("Number of GO terms for ontology", ontology, ":", nrow(go_df)))  # Debugging
  
  # Filter significant results
  filtered_go <- go_df %>%
    filter(p.adjust < 0.05) %>%
    dplyr::select(Description, Cluster, p.adjust, GeneRatio)
  
  print(paste("Number of significant GO terms for ontology", ontology, ":", nrow(filtered_go)))  # Debugging
  
  # If there are no significant terms, stop the function here
  if (nrow(filtered_go) == 0) {
    print(paste("No significant GO terms for ontology", ontology))
    return(NULL)  # Exit if no significant results
  }
  
  # Separate by Cluster (FL vs trunc) and examine the differences
  fl_terms <- filtered_go %>% filter(Cluster == "FL")
  trunc_terms <- filtered_go %>% filter(Cluster == "trunc")
  
  # Identify the GO terms that are different (unique for each isoform)
  fl_unique <- setdiff(fl_terms$Description, trunc_terms$Description)
  trunc_unique <- setdiff(trunc_terms$Description, fl_terms$Description)
  
  # Select top N unique terms by significance (lowest p.adjust)

  top_n <- 15

  fl_top <- fl_terms %>%
  filter(Description %in% fl_unique) %>%
  arrange(p.adjust) %>%
  slice_head(n = top_n)

  trunc_top <- trunc_terms %>%
  filter(Description %in% trunc_unique) %>%
  arrange(p.adjust) %>%
  slice_head(n = top_n)

  
top_unique_terms <- bind_rows(
  fl_top %>% mutate(Cluster = "TrkB.FL"),
  trunc_top %>% mutate(Cluster = "TrkB.T1")
)
top_unique_terms$GeneRatio <- sapply(strsplit(top_unique_terms$GeneRatio, "/"), function(x) as.numeric(x[1]) / as.numeric(x[2]))
top_unique_terms$Description <- str_wrap(top_unique_terms$Description, width = 40)

 top_unique_terms$ShortDescription <- sapply(strsplit(top_unique_terms$Description, " "), function(words) {
  paste(head(words, 4), collapse = " ")
})
  # If there are no unique terms, stop the function here
  if (nrow(top_unique_terms) == 0) {
    print(paste("No unique GO terms for ontology", ontology))
    return(NULL)  # Exit if no unique terms
  }
  


  # Plot using ggplot
  p <- ggplot(top_unique_terms, aes(x  = Cluster , y = reorder(ShortDescription, p.adjust))) +
  geom_point(aes(size = -log10(p.adjust), color = GeneRatio)) +
  coord_flip() +
  labs(
    title = paste("Top Unique GO Terms:", ontology_full, "- TrkB.FL vs TrkB.T1"),
    x = "Transcript Isoform", y = "Go Term",
    size = "-log10(p.adjust)", color = "GeneRatio"
  ) +
  scale_color_viridis_c(option = "C", direction = 1) +
  scale_x_discrete(expand = expansion(mult = c(0.4, 0.4))) +
  theme_minimal() +
  theme(
  
    plot.title = element_text(hjust = 0),
    plot.title.position = "plot",
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    axis.text.y = element_text(size = 10),
    plot.margin = margin(t = 10, r = 10, b = 20, l = 15),  # top, right, bottom, left
    legend.position = "right",
    legend.box.margin = margin(0, 10, 0, 0)
  )
  print(p)
  
  ggsave(
  filename = paste0("GO_TopUnique_", ontology_full, ".png"),
  plot = p,
  width = 10,
  height = 6,
  dpi = 300
)
}
# Generate plots for each ontology
for (ontology in ontologies) {
  generate_plots_for_ontology(ontology)
}

```

# M10 Seurat

*Processing M10 originated Seurat*
```{r}
# Define the directory containing RDS files
rds_directory_orig <- "/data/gpfs/projects/punim2183/samples_processing/count_matrix_files/orig"

# 1. Load with renaming data
seurat_list_M10 <- read_rds_to_seurat(rds_directory_orig) |>
  rename_cells_in_combined_seurat_bothects()

# 2. Normalize with SCTransform and merge
combined_seurat_M10 <- normalize_and_merge(seurat_list_M10) |>
  update_orig_ident()

# 3. Process with Harmony integration
combined_seurat_M10 <- process_seurat(combined_seurat_M10)
combined_seurat_M10 <- run_harmony(combined_seurat_M10)

# 4. Save the processed object
saveRDS(combined_seurat_M10, "combined_seurat_M10.rds")


# Print a success message
print("Workflow completed successfully! The combined Seurat object has been saved.")

```

```{r}

#saving files and loading the previous step to save computational time
#combined_seurat_both <- readRDS("/data/gpfs/projects/punim2183/data_processed/combined_seurat_both.rds")

names(combined_seurat_M10@reductions)

combined_seurat_M10 <- JoinLayers(combined_seurat_M10, assay = "RNA")

```

```{r, fig.width=12, fig.height=10, echo=FALSE}
# combined_seurat_M10 <- readRDS("/data/gpfs/projects/punim2183/data_processed/combined_seurat_M10.rds")

neurons_markers = c("Map2", "Rbfox3", "Tubb3")
oligos_markers = c("Mbp", "Olig1", "Olig2")
astrocytes_markers = c("Gfap", "Aqp4", "Slc1a3")

count_expressing_cells_M10 <- function(markers, combined_seurat_M10) {
  expression_data <- FetchData(combined_seurat_M10, vars = markers)
  num_cells <- sum(rowSums(expression_data > 0) > 0)
  return(num_cells)
}

markers_feature_plot_M10 <- FeaturePlot(combined_seurat_M10, c(neurons_markers, astrocytes_markers, oligos_markers))
markers_feature_plot_M10

dimplot_clusters_M10 <- DimPlot(combined_seurat_M10, reduction = "umap", group.by = "seurat_clusters", label = TRUE) 
dimplot_clusters_M10

ggsave("markers_feature_plot_M10.png", plot = markers_feature_plot_M10, width = 10, height = 8, dpi = 300)
ggsave("dimplot_clusters_M10.png", plot = dimplot_clusters_M10, width = 10, height = 8, dpi = 300)


```

```{r}

cluster_to_celltype <- list(
  Neurons = c(5, 3, 20, 28, 19, 11, 14),
  Oligo = c(10, 16, 17, 31, 7, 27, 29),
  Astrocytes = c(1, 2, 8, 18, 9, 26, 23, 29)
)

# Create a vector to store cell type labels
cell_type_labels_M10 <- rep(NA, ncol(combined_seurat_M10))

# Assign specific cell types based on cluster IDs
for (cluster_id in 0:31) {
  if (cluster_id %in% cluster_to_celltype$Neurons) {
    cell_type_labels_M10[combined_seurat_M10$seurat_clusters == cluster_id] <- "Neurons"
  } else if (cluster_id %in% cluster_to_celltype$Oligo) {
    cell_type_labels_M10[combined_seurat_M10$seurat_clusters == cluster_id] <- "Oligo"
  } else if (cluster_id %in% cluster_to_celltype$Astrocytes) {
    cell_type_labels_M10[combined_seurat_M10$seurat_clusters == cluster_id] <- "Astrocytes"
  }
}

# Assign generic labels to remaining clusters
remaining_clusters_M10 <- setdiff(0:31, unlist(cluster_to_celltype))
for (i in seq_along(remaining_clusters_M10)) {
  cluster_id <- remaining_clusters_M10[i]
  cell_type_labels_M10[combined_seurat_M10$seurat_clusters == cluster_id] <- paste0("cell_type", i)
}


# Add the cell type labels to the Seurat object
combined_seurat_M10$cell_type <- cell_type_labels_M10

# Set the cell type as the active identity
Idents(combined_seurat_M10) <- combined_seurat_M10$cell_type
table(combined_seurat_M10$cell_type)

dimplot_cell_types_M10 <- DimPlot(combined_seurat_M10, reduction = "umap", label = TRUE, group.by = "cell_type", raster = FALSE)
dimplot_cell_types_M10
ggsave("dimplot_cell_types_M10.jpg", plot= dimplot_cell_types_M10,width = 10, height = 8, dpi = 300)

#check metadata
M10_metadata <- combined_seurat_M10@meta.data
head(M10_metadata)


```

## Subset seurat to cell types of interest for focused plotting 
```{r, echo=FALSE}
cell_types_of_interest <- c("Neurons", "Oligo", "Astrocytes")

subset_seurat_M10 <- subset(combined_seurat_M10, subset = cell_type %in% cell_types_of_interest)


# Generate Plots
vln_plot_M10 <- VlnPlot(subset_seurat_M10, features = "Ntrk2", pt.size = 0)
feature_plot_M10 <- FeaturePlot(subset_seurat_M10, features = "Ntrk2", raster = FALSE)
dim_plot_M10 <- DimPlot(subset_seurat_M10)
vln_plot_M10
feature_plot_M10
dim_plot_M10

ggsave("feature_plot_M10.png", plot = feature_plot_M10, width = 10, height = 8, dpi = 300)
ggsave("dim_plot_M10.png", plot = dim_plot_M10, width = 10, height = 8, dpi = 300)
ggsave("vln_plot_M10.png", plot = vln_plot_M10, width = 10, height = 8, dpi = 300)

#saveRDS(subset_seurat_M10, "subset_seurat_M10.rds")
```

```

