    library(SeuratObject)
    library(SeuratDisk)
    library(Seurat)
    library(cowplot)
    library(dplyr)
    library(ggplot2)
    library(readr)
    library(DT)
    library(tidyr)
    library(patchwork)
    library(harmony)
    # install.packages("RCurl")
    # BiocManager::install("rtracklayer")

    library(rtracklayer)

    source("/data/gpfs/projects/punim2183/data_processed/dropEst_analysis_functions.R")
    #save.image(file = "dropEst_output.RData")
    #load("dropEst_output.RData")

    # knitr::opts_chunk$set(eval = FALSE)
    # rmarkdown::render("dropEst_output.Rmd", output_format = "md_document", output_file = "dropEst_output_analysis.md")

\##Step 1: Reading count matrices from a directory and createing Seurat
objects:

For each count matrix type (trunc, FL, orig and both) generated by
dropest using different gtf file.

    # Define the directory containing RDS files
    rds_directory_FL <- "/data/gpfs/projects/punim2183/samples_processing/count_matrix_files/FL"

    # Step 1: Read RDS files into Seurat objects
    seurat_list_FL <- read_rds_to_seurat(rds_directory_FL)

# Step 2: Rename cells in the Seurat objects:

Extract naming patterns from the names of Seurat objects. And rename
cell names in each Seurat object using the extracted naming patterns.
Then combine multiple Seurat objects into a single Seurat object.

### Extract the naming pattern (e.g., “10X05\_1”)

**input** seurat\_objects\_trunc (a list of Seurat objects with names
like count\_matrix.rds\_10X05\_1.bam.1\_trunc.gtf).

**Output** Each Seurat object in seurat\_objects\_trunc now has cell
names prefixed with the corresponding naming pattern (e.g.,
“10X05\_1\_Cell\_1”).

    seurat_list_FL <- rename_cells_in_seurat_objects(seurat_list_FL)

# Step 3: Normalize and merge Seurat objects

Preprocess the combined Seurat object by normalizing, scaling,
clustering, and filtering the data. Then save the final Seurat object
into an rds file format.

**Input:** List of seurat objects .

**Output:** A Normalised and merged Seurat object with:

    combined_seurat_FL <- normalize_and_merge(seurat_list_FL)

# Step 4: Update orig.ident metadata

    combined_seurat_FL <- update_orig_ident(combined_seurat_FL)

# Step 5: Process the merged Seurat object

Process the combined Seurat object by scaling, clustering, and filtering
the data. **Input:** combined\_seurat\_trunc (the combined Seurat
object).

**Output:** A processed Seurat object with:

-   scaled data.

-   PCA, clustering, and UMAP performed.

-   Cells filtered based on mitochondrial content (percent.mt &lt; 5),
    feature count (nFeature\_RNA &gt; 200 & nFeature\_RNA &lt; 2500),
    and RNA count.

<!-- -->

    combined_seurat_FL <- process_seurat(combined_seurat_FL)

# Step 6: Integrate datasets using Harmony

For addressing batch effect, grouped by orig.ident

    combined_seurat_FL <- run_harmony(combined_seurat_FL)

# Step 7: Save the final Seurat object

    save_seurat_object(combined_seurat_FL, "combined_seurat_FL.rds")

    # Print a success message
    print("Workflow completed successfully! The combined Seurat object has been saved.")

    #seurat_objects_orig <- read_rds_to_seurat("/data/gpfs/projects/punim2183/samples_processing/count_matrix_files/orig")

    #seurat_objects_trunc <- read_rds_to_seurat("/data/gpfs/projects/punim2183/samples_processing/count_matrix_files/trunc")

    #seurat_objects_FL <- read_rds_to_seurat("/data/gpfs/projects/punim2183/samples_processing/count_matrix_files/FL")

Extract naming patterns from the names of Seurat objects. And rename
cell names in each Seurat object using the extracted naming patterns.
Then combine multiple Seurat objects into a single Seurat object.

### Extract the naming pattern (e.g., “10X05\_1”)

*input:* seurat\_objects\_trunc (a list of Seurat objects with names
like count\_matrix.rds\_10X05\_1.bam.1\_trunc.gtf).

*Output:* Each Seurat object in seurat\_objects\_trunc now has cell
names prefixed with the corresponding naming pattern (e.g.,
“10X05\_1\_Cell\_1”).

    names_list_trunc <- names(seurat_objects_trunc)

    naming_patterns_trunc <- gsub("^.*_(10X\\d+_\\d+)\\..*$", "\\1", names_list_trunc)

    # Print the naming patterns
    print(naming_patterns_trunc)

    for (i in 1:length(seurat_objects_trunc)) {
        prefix_trunc <- naming_patterns_trunc[i]  

        current_cell_names_trunc <- colnames(seurat_objects_trunc[[i]])  

        new_cell_names_trunc <- paste0(prefix_trunc, "_", current_cell_names_trunc)  

        colnames(seurat_objects_trunc[[i]]) <- new_cell_names_trunc
    }

### Preprocessing and Merging Seurat objects:

first, normalise each Seurat object, ensures each dataset is adjusted
for technical differences independently, reducing potential batch
effects.

Addressing batch effect: some samples had 22k cells while others had 3k,
6k .. which is done by using Harmony package

(Normalizing before merging ensures that each sample is normalized
independently, which is important if there are significant differences
in sequencing depth between samples.

*Input:* seurat\_objects\_trunc (list of Seurat objects).

*Output:* Normalised one Seurat object, preprocessed (scaling, PCA), use
of Harmony integration for Clustering and umap embedding.

    # Normalize each dataset separately
    seurat_objects_trunc <- lapply(seurat_objects_trunc, NormalizeData)

    # Merge the datasets, keeping normalized data
    combined_seurat_trunc <- merge(
      seurat_objects_trunc[[1]], 
      y = seurat_objects_trunc[-1], 
      merge.data = TRUE
    )

    #update orig.ident
    sample_names <- sub("^(10X[0-9]+_[0-9]+)_.*", "\\1", colnames(combined_seurat_trunc))
    combined_seurat_trunc$orig.ident <- sample_names

    head(combined_seurat_trunc$orig.ident)
    table(combined_seurat_trunc$orig.ident)

    # Find variable features
    combined_seurat_trunc <- FindVariableFeatures(combined_seurat_trunc)

    # Scale and run PCA
    combined_seurat_trunc <- ScaleData(combined_seurat_trunc)
    combined_seurat_trunc <- RunPCA(combined_seurat_trunc)

    # Run Harmony for integration
    combined_seurat_trunc <- RunHarmony(
      object = combined_seurat_trunc,
      group.by.vars = "orig.ident"  
    )

    ElbowPlot(combined_seurat_trunc, ndims = 50)  

    # Use Harmony embeddings for clustering & visualization
    combined_seurat_trunc[["percent.mt"]] <- PercentageFeatureSet(combined_seurat_trunc, pattern = "^MT-")
    combined_seurat_trunc <- subset(combined_seurat_trunc,
        subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5
      )
    combined_seurat_trunc <- FindNeighbors(combined_seurat_trunc, reduction = "harmony")
    combined_seurat_trunc <- FindClusters(combined_seurat_trunc, resolution = 0.5) # moderate resolution 
    combined_seurat_trunc <- RunUMAP(combined_seurat_trunc, reduction = "harmony", dims = 1:30)

    saveRDS(combined_seurat_trunc, file = "combined_seurat_object_trunc.rds")

    install.packages("clustree")
    library(clustree)
    seurat_object <- FindClusters(combined_seurat_trunc, resolution = seq(0.2, 1.0, by = 0.2))
    clustree(seurat_object, prefix = "RNA_snn_res.")

Preprocess the combined Seurat object by normalizing, scaling,
clustering, and filtering the data. Then save the final Seurat object
into an rds file format.

*Input:* combined\_seurat\_trunc (the combined Seurat object).

*Output:* A preprocessed (and saved) Seurat object with:

\*scaled data.

\*PCA, clustering, and UMAP performed.

\*Cells filtered based on mitochondrial content (percent.mt &lt; 5),
feature count (nFeature\_RNA &gt; 200 & nFeature\_RNA &lt; 2500), and
RNA count.

    saveRDS(combined_seurat_trunc, file = "combined_seurat_object_trunc.rds")


    # Bar plot of cell counts by sample
    barplot(table(combined_seurat_trunc$orig.ident), 
            las = 2,  # Rotate x-axis labels
            main = "Cell Counts by Sample", 
            xlab = "Sample", 
            ylab = "Number of Cells")

## Gene expression analysis:

    gene_expression_Ntrk2_trunc <- FetchData(combined_seurat_trunc, vars = "Ntrk2trunc")
    total_expression_Ntrk2_trunc <- sum(gene_expression_Ntrk2_trunc$"Ntrk2trunc")
    total_expression_Ntrk2_trunc

    FeaturePlot(combined_seurat_trunc, features = "Ntrk2trunc")

    neurons_markers = c("Map2", "Rbfox3", "Tubb3")
    oligos_markers = c("Mbp", "Olig1", "Olig2")
    astrocytes_markers = c("Gfap", "Aqp4", "Slc1a3")

    count_expressing_cells <- function(markers, combined_seurat_trunc) {
      expression_data <- FetchData(combined_seurat_trunc, vars = markers)
      num_cells <- sum(rowSums(expression_data > 0) > 0)
      return(num_cells)
    }

    # Count cells expressing each set of markers
    neurons_cells_expressing <- count_expressing_cells(neurons_markers, combined_seurat_trunc)
    oligos_cells_expressing <- count_expressing_cells(oligos_markers, combined_seurat_trunc)
    astrocytes_cells_expressing <- count_expressing_cells(astrocytes_markers, combined_seurat_trunc)

    print(paste("Number of cells expressing neuron markers:", neurons_cells_expressing))
    print(paste("Number of cells expressing oligodendrocyte markers:", oligos_cells_expressing))
    print(paste("Number of cells expressing astrocyte markers:", astrocytes_cells_expressing))

    FeaturePlot(combined_seurat_trunc, features = c("Map2", "Rbfox3", "Tubb3", "Mbp", "Olig1", "Olig2", "Gfap", "Aqp4", "Slc1a3"))
    FeaturePlot(combined_seurat_trunc, features = "Ntrk2trunc")
    #FeaturePlot(combined_seurat_FL, features = "Ntrk2FL")
    # FeaturePlot(combined_seurat_orig, features = "Ntrk2")
    DimPlot(combined_seurat_trunc)

    gene_names <- c("Ntrk2trunc", "Ntrk2FL")  
    gene_expression <- FetchData(seurat_obj_1, vars = gene_names)

    num_cells_expressing <- list()

    for (gene in gene_names) {
      num_cells_expressing[[gene]] <- sum(gene_expression[[gene]] > 0)
      print(paste("Number of cells expressing", gene, ":", num_cells_expressing[[gene]]))
    }

    #cluster_markers <- FindAllMarkers(seurat_obj_1, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

    # cell type annotation
    new_cluster_ids <- c("Astrocytes", "Cell_type_1", "Neurons", "Oligo", "Cell_type_2", "Cell_type_3", "Oligo", "Cell_type_4", "Cell_type_5", "Cell_type_6")
    names(new_cluster_ids) <- levels(seurat_obj_1)
    seurat_obj_1 <- RenameIdents(seurat_obj_1, new_cluster_ids)
    Idents(seurat_obj_1)

    # seurat_obj_1 <- SetIdent(seurat_obj_1, value = "seurat_clusters")


    FeaturePlot(seurat_obj_1, features = neurons_markers)
    FeaturePlot(seurat_obj_1, features = oligos_markers)
    FeaturePlot(seurat_obj_1, features = astrocytes_markers)


    VlnPlot(seurat_obj_1, features = neurons_markers, group.by = "seurat_clusters")
    VlnPlot(seurat_obj_1, features = astrocytes_markers, group.by = "seurat_clusters")
    VlnPlot(seurat_obj_1, features = oligos_markers, group.by = "seurat_clusters")



    DotPlot(seurat_obj_1, features = c(neurons_markers, oligos_markers, astrocytes_markers)) + RotatedAxis()

    # subset the Seurat object to include only cell_types 
    # subset_seurat <- subset(seurat_obj_1, idents = c("Astrocytes", "Neurons", "Oligo"))
                         

    # VlnPlot(subset_seurat, features = "Ntrk2trunc")
    # VlnPlot(subset_seurat, features = "Ntrk2FL")
    VlnPlot(subset_seurat, features = gene_names)
    expression_data <- FetchData(subset_seurat, vars = gene_names)


    expression_summary <- expression_data %>%
      group_by(Idents(subset_seurat)) %>%
      summarise(across(everything(), mean, na.rm = TRUE))

    print(expression_summary)

    # DotPlot(subset_seurat, features = c("Ntrk2trunc", "Ntrk2FL"))

    # Install the SeuratData package if needed
    if (!requireNamespace("SeuratData", quietly = TRUE)) {
        install.packages("SeuratData")
    }

    # Load SeuratData package
    library(SeuratData)

    # Install the stxBrain dataset
    SeuratData::InstallData("stxBrain")


    options(timeout=600)
    InstallData("stxBrain")
    brain <- LoadData("stxBrain", type = "anterior1")
    VlnPlot(brain, features = "nCount_Spatial", pt.size = 0.1, layer = "counts") + NoLegend()
    SpatialPlot(brain, features = "nCount_Spatial", slot = "counts", image.alpha = 1)

    # Normalization 
    brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)  # For SCTransform normalization

    #log1p
    cMat <- brain[["Spatial"]]$counts
    cMat <- log1p(cMat)
    brain[["log1p"]] <- CreateAssayObject(
      data = cMat
    )

    SpatialFeaturePlot(brain, features = "Ntrk2")
